mozilla/blurts-server
# MNTOR-3029 - guard against un-parseable URLs being passed to next-auth redirect

# References:

Jira: MNTOR-3029



# Description

Currently next-auth throws an unhandled exception if there are fuzzing attempts against the `redirect` endpoint. We do want to log these, but it being unhandled means it pollutes Sentry and our logs with a lot of useless information.

This modifies the default behavior (per the documentation) to:

1. use the `URL.canParse` static method to try parsing any incoming URLs
2. if it cannot be parsed, log this and return `baseUrl`

We should consider trying to upstream this to the next-auth project, I think it's a better default but one problem is that logging is not very standardized in the JS ecosystem so I'm not sure what a good upstream fix would look like yet.

NOTE: this introduces a slight change in behavior - the way this feature works is that the `next-auth.callback-url` cookie is set to whatever the URL passed in a `POST` to `/api/auth/signin/fxa` via the`callbackUrl` form field is. If the `callbackUrl` starts with a `/` then it is appended to `baseUrl` (relative), otherwise it's treated as an absolute URL.

The URL in the cookie is URL-encoded, but other than that it seems to allow pretty much any path here. For absolute URLs there's an origin check using `new URL(...).origin`, and relative URLs must begin with `/` and always prepend the `baseUrl`.

Before this PR, next-auth throws an unhandled exception from `new URL()` so HTTP 500 is returned to the caller. With this PR, any un-pareseable URLs are logged and `baseUrl` is returned, so the caller gets an HTTP 200 instead.

I'm going to land this after the next release so we have more time for testing, and I also want to file an upstream issue with next-auth because I think this should be better handled by next-auth itself. It shouldn't be possible for callers to just cause an HTTP 500 to be thrown through invalid input - this is a client error and should be HTTP 4xx class (probably `400 Bad Request`)

# How to test

Sign in/out should work as normal and redirect to any valid URL. Attempting to sign in with an invalid redirect URL should be ignored and the request should be redirected to `/`.

See the Jira ticket for an example test case that triggers the HTTP 500 response from next-auth without this PR.

# Checklist (Definition of Done)

- [ ] File a ticket on next-auth project
